PR: Add search paths to PATH in is_program_installed

mrclary opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes
Temporarily add required paths to `sys.environ['PATH']` in `spyder.utils.programs.is_program_installed`.


* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

<!--- Explain what you've done and why --->
By adding platform dependent paths to `os.environ['PATH']`, developers will not have to consider macOS application cases when using `find_program`. Additionally, developers will not have to consider cases in which Spyder is not launched from login shells (i.e. user profiles are not sourced).

Since `PATH` is only temporarily modified to locate a given program, it should not affect the behavior of Spyder or Consoles.

Developers should update the `req_paths` in the respective platform control sequence to include paths in the search for programs.

### Issues Resolved
Fixes #14136 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct:
mrclary
<!--- Thanks for your help making Spyder better for everyone! --->

mrclary commented:
@ccordoba12 @steff456, per conversation [here](https://github.com/spyder-ide/spyder/pull/13950#discussion_r516201754), after merging #13950, this PR can be rebased and reviewed.

jitseniesen commented:
Why is it necessary to change the environment variable PATH at all? Would it be possible instead to change
```
for path in os.environ["PATH"].split(os.pathsep):
```
to
```
for path in os.environ["PATH"].split(os.pathsep) + req_paths:
```
or something like that?

mrclary commented:
> Why is it necessary to change the environment variable PATH at all? Would it be possible instead to change
> 
> ```
> for path in os.environ["PATH"].split(os.pathsep):
> ```
> 
> to
> 
> ```
> for path in os.environ["PATH"].split(os.pathsep) + req_paths:
> ```
> 
> or something like that?

@jitseniesen, yours is a much better idea. Thanks.

ccordoba12 commented on spyder/utils/programs.py:
```diff
@@ -71,17 +71,50 @@ def get_temp_dir(suffix=None):
+            osp.join('/opt', 'miniconda3', 'condabin')
+        ])
+    elif sys.platform.startswith('linux'):
+        # TODO: what are the possible conda paths?
+        # TODO: what are the possible pyenv paths?
+        pass
```
```suggestion
        req_paths.extend([
            osp.join('/usr', 'local', 'bin'),
            osp.join(home, 'anaconda3', 'condabin'),
            osp.join(home, 'miniconda3', 'condabin'),
            osp.join('/opt', 'anaconda3', 'condabin'),
            osp.join('/opt', 'miniconda3', 'condabin')
        ])
```

ccordoba12 commented on spyder/utils/programs.py:
```diff
@@ -71,17 +71,50 @@ def get_temp_dir(suffix=None):
+    launch Spyder.
+
+    On macOS systems, a .app is considered installed if it exists.
+
+    Darwin:
+        /usr/local/bin
+        ~/opt/anaconda3/condabin
+        ~/opt/miniconda3/condabin
+        /opt/anaconda3/condabin
+        /opt/miniconda3/condabin
+    Linux:
+        ?
+    Windows:
+        ?
```
```suggestion
```

Default paths are present in the code below, so I see no need to add them here.

ccordoba12 commented on spyder/utils/programs.py:
```diff
@@ -71,17 +71,50 @@ def get_temp_dir(suffix=None):
+        # TODO: what are the possible pyenv paths?
+        pass
+    elif WINDOWS:
+        # TODO: what are the possible conda paths?
+        # TODO: what are the possible pyenv paths?
+        pass
```
@dalthviz, could you provide the default installation locations for Anaconda and Miniconda on Windows here?

dalthviz replied:
For Anaconda:

* Installations for just the current user: the path is something like `C:\Users\<user>\anaconda3\condabin`
* Installations for all the user: the default path is something like `C:\ProgramData\Anaconda3\condabin`

For Miniconda:

* Installations for just the current user: the path is something like `C:\Users\<user>\miniconda3\condabin`
* Installations for all the user: the default path is something like `C:\ProgramData\Miniconda3\condabin`

mrclary replied:
Looks like standard installation location for `pyenv` is `$HOME\.pyenv\pyenv-win\bin` and for `conda` is 
`$HOME\Anaconda3\condabin`
`$HOME\Miniconda3\condabin`
`C:\Anaconda3\condabin`
`C:\Miniconda3\condabin`
`C:\ProgramData\Anaconda3\condabin`
`C:\ProgramData\Miniconda3\condabin`

@dalthviz, can you confirm? Additions? Subtractions?

dalthviz replied:
LGTM :+1:

ccordoba12 reviewed (commented):
Thanks @mrclary, this looks pretty good. I left an initial review for you.

ccordoba12 reviewed (approved):
Looks good to me now, thanks @mrclary!
